compiler/rustc_trait_selection/src/traits/fulfill.rs

```diff
@@ -355,6 +357,21 @@ impl<'a, 'b, 'tcx> FulfillProcessor<'a, 'b, 'tcx> {
 
         let infcx = self.selcx.infcx();
 
+        if obligation.predicate.has_projections() {
+            let mut obligations = Vec::new();
+            let predicate = crate::traits::project::try_normalize_with_depth_to(
+                self.selcx,
+                obligation.param_env,
+                obligation.cause.clone(),
+                obligation.recursion_depth + 1,
+                obligation.predicate,
+                &mut obligations,
+            );
+            if predicate != obligation.predicate {
```

Comment:
another optimization here might be to not `return` if `obligations.is_empty()`

Reply:
Yeah, but if that's the only obligation (and we have no nested obligations that we need to check due to the normalization), then we can just update `obligation.predicate = predicate` like the shallow-resolve code below.